runtime/tools/python/ttrt/common/run.py

```diff
@@ -563,6 +593,65 @@ def _execute(binaries):
+                                    num_items = 0
+                                    for (
+                                        key,
+                                        value,
+                                    ) in callback_runtime_config.memory_report.items():
+                                        num_items += 1
+
+                                    if num_items == 0:
+                                        self.logging.warning(f"No memory data found")
+                                    else:
+                                        # query initial memory usage
+                                        dram_initial_size = callback_runtime_config.memory_report[
+                                            0
+                                        ][
+                                            "dram"
+                                        ][
+                                            "total_allocated (bytes) : total_allocated/bank * num_banks"
+                                        ]
+                                        l1_initlal_size = callback_runtime_config.memory_report[
+                                            0
+                                        ][
+                                            "l1"
+                                        ][
+                                            "total_allocated (bytes) : total_allocated/bank * num_banks"
+                                        ]
+
+                                        # query final memory usage and ensure no memory leaks
+                                        dram_final_size = callback_runtime_config.memory_report[
+                                            num_items - 1
+                                        ][
+                                            "dram"
+                                        ][
+                                            "total_allocated (bytes) : total_allocated/bank * num_banks"
+                                        ]
+                                        l1_final_size = callback_runtime_config.memory_report[
+                                            num_items - 1
+                                        ][
+                                            "l1"
+                                        ][
+                                            "total_allocated (bytes) : total_allocated/bank * num_banks"
+                                        ]
+
+                                        if dram_final_size > dram_initial_size:
+                                            raise Exception(
+                                                "Memory leak detected in DRAM"
+                                            )
+
+                                        if l1_final_size > l1_initlal_size:
+                                            raise Exception(
+                                                "Memory leak detected in L1 cache"
+                                            )
+
```

Comment:
** This function is getting a bit unwieldy esp with nested control flow.  This new code might be better encapsulated inside of a utility function.